### PR TITLE
TheMediaGrid Bid Adapter: fill user.id from fpd cookie

### PR DESCRIPTION
**Pull request by TheMediaGrid**
## Type of change

- [x] Feature

## Description of change

Check fpd cookie and send it in ad request if available

- grid-tech@themediagrid.com
- [x] official adapter submission

### CODE REVIEW

**Comment by patmmccann** [comment]
this appears to largely duplicate the pubcid and I believe should not be allowed.

**Comment by patmmccann** [comment]
@TheMediaGrid I opened issue https://github.com/prebid/Prebid.js/issues/7280 to discuss if this functionality is allowed.

**Comment by patmmccann** [comment]
Hi @TheMediaGrid ; we discussed this at length in the js committee today: Can you make sure all this behavior is opt-in by a publisher?

 Also, we plan to make it so publishers can prevent adapters from using the storage.setCookie method in the future, but that functionality is not yet available.

File: modules/gridBidAdapter.js [hunk 1]
```diff
@@ -427,6 +437,18 @@ function createBannerRequest(bid, mediaType) {
   return result;
 }
 
+function makeNewUserIdInFPDCookie() {
```
Comment (patmmccann) [inline modules/gridBidAdapter.js:440]: please make this dependent on adapter config that defaults to 'off'

**Review by patmmccann — CHANGES_REQUESTED** [review]

**Comment by patmmccann** [comment]
Hello @TheMediaGrid 

Any feedback on the requested changes?

**Comment by TheMediaGrid** [comment]
Hello! We have added bidder config option 'cookieSettingAllowed' to control cookies adding

File: modules/gridBidAdapter.js [hunk 2]
```diff
@@ -427,6 +437,21 @@ function createBannerRequest(bid, mediaType) {
...
+  if (config.getConfig('cookieSettingAllowed')) {
+    const value = utils.generateUUID().replace(/-/g, '');
+    const expires = new Date(Date.now() + USER_ID_COOKIE_EXP).toISOString();
+
+    storage.setCookie(USER_ID_KEY, value, expires);
```
Comment (patmmccann) [inline modules/gridBidAdapter.js:445]: why not use local storage?

File: modules/gridBidAdapter.js [hunk 3]
```diff
@@ -427,6 +437,21 @@ function createBannerRequest(bid, mediaType) {
...
+    return value;
+  }
+  return null;
+}
+
+function getUserIdFromFPDCookie() {
+  return storage.getCookie(USER_ID_KEY) || makeNewUserIdInFPDCookie();
```
Comment (patmmccann) [inline modules/gridBidAdapter.js:452]: why not use local storage?

File: modules/gridBidAdapter.js (hunk 2 above)
Comment (patmmccann) [reply modules/gridBidAdapter.js:445]: Expanding on this: cookies get appended to all requests that a browser makes to the publisher in the http headers, and can lead to overflows of CDN limits and failed requests. Eg cloudflare has a 4k limit. Publishers that make tons of requests to the same domain may see actual bandwidth consumed noticibly change. Local storage does not have this disadvantage.

**Comment by TheMediaGrid** [comment]
Hello! Now we use localStorage instead Cookies

**Comment by TheMediaGrid** [comment]
@patmmccann Hello, is this ready to merge, or there is some additional questions?

**Comment by patmmccann** [comment]
> @patmmccann Hello, is this ready to merge, or there is some additional questions?

Please add a docs pr, code lgtm

**Comment by TheMediaGrid** [comment]
Pull request in the doc repo - https://github.com/prebid/prebid.github.io/pull/3254

**Review by patmmccann — APPROVED** [review]

**Comment by patmmccann** [comment]
Merging, but please note my change request on the docs PR as well as please rebase to master your other two pr's on your adapter to these changes don't show. Thanks!

**Pull request merged**